src/Compilers/CSharp/Portable/Operations/CSharpConversionExpression.cs

```diff
@@ -0,0 +1,67 @@
+    {
+        protected BaseCSharpConversionExpression(Conversion conversion, bool isExplicitInCode, bool throwsExceptionOnFailure, bool isChecked, SyntaxNode syntax, ITypeSymbol type, Optional<object> constantValue) :
+            base(conversion, isExplicitInCode, throwsExceptionOnFailure, isChecked, syntax, type, constantValue)
+        {
+            ConversionInternal = conversion;
+        }
+
+        internal Conversion ConversionInternal { get; }
+
+        public override string LanguageName => LanguageNames.CSharp;
+    }
+
+    internal sealed partial class CSharpConversionExpression : BaseCSharpConversionExpression
+    {
+        public CSharpConversionExpression(IOperation operand, Conversion conversion, bool isExplicit, bool throwsExceptionOnFailure, bool isChecked, SyntaxNode syntax, ITypeSymbol type, Optional<object> constantValue) :
+            base(conversion, isExplicit, throwsExceptionOnFailure, isChecked, syntax, type, constantValue)
+        {
+            Operand = operand;
+        }
+
+        public override IOperation Operand { get; }
+    }
+
+    internal sealed partial class LazyCSharpConversionExpression : BaseCSharpConversionExpression
+    {
+        private readonly Lazy<IOperation> _operand;
+        public LazyCSharpConversionExpression(Lazy<IOperation> operand, Conversion conversion, bool isExplicit, bool throwsExceptionOnFailure, bool isChecked, SyntaxNode syntax, ITypeSymbol type, Optional<object> constantValue) :
+            base(conversion, isExplicit, throwsExceptionOnFailure, isChecked, syntax, type, constantValue)
+        {
+            _operand = operand;
+        }
+
+        public override IOperation Operand => _operand.Value;
+    }
+
+    public static class IConversionExpressionExtensions
+    {
+        /// <summary>
+        /// Gets the underlying <see cref="Conversion"/> information from this <see cref="IConversionExpression"/>. This
+        /// <see cref="IConversionExpression"/> must have been created from CSharp code.
+        /// </summary>
+        /// <param name="conversionExpression">The conversion expression to get original info from.</param>
+        /// <returns>The underlying <see cref="Conversion"/>.</returns>
+        /// <exception cref="InvalidCastException">If the <see cref="IConversionExpression"/> was not created from CSharp code.</exception>
+        public static Conversion GetConversion(this IConversionExpression conversionExpression)
+        {
+            if (conversionExpression is BaseCSharpConversionExpression csharpConversionExpression)
+            {
+                return csharpConversionExpression.ConversionInternal;
+            }
+            else
+            {
+                throw new ArgumentException(string.Format(CSharpResources.IConversionExpressionIsNotCSharpConversion,
+                                                          nameof(IConversionExpression),
+                                                          conversionExpression));
```

Comment:
I haven't updated the tests at all for the new conversion scheme, so this method hasn't been tested at all yet. We are currently relying on `ToString`, and there probably isn't a reasonable string representation currently. Is our convention to include the object in the error message, or to not include it and just say `IConversionExpression is not a C#/VB conversion`?